### PR TITLE
Cutomisable Cash Flow Reports

### LINKED ISSUE #8109: Incomplete Mapping in Cashflow
Dear Team,

Much work has been done and kudos.

I just discovered that not all elements that constitute a standard cashflow report is linked to cashflow report in erpnext.

Examples are:
1. Fixed Asset disposal. The gain or loss is not adjusted in the cash flow.
2. If company income tax is paid.The amount paid is not adjusted in the cash flow.
3. No Adjustment for deferred tax
4. For investing activities only net change in fixed assets is accommodated. Other items like Investment purchase and disposal are not linked.
5. For financing activities,only change in equity is allowed,other financing options are not linked.
6. Cash activities relating to the following account groups in the chart of accounts does reflect in the cashflow: Loan and Advances (assets),Security and Deposit,Tax Assets,Investmnet,Duties and Taxes,Loan (Liability)

Kindly please link the necessary field in the cashflow report.

Here is a link for a standard cashflow:
http://www.excel-skills.com/templates/cashflow_statement_sample.xls

### PR DESCRIPTION
**Pull request by tundebabzy**
Closes #8109 

As rightly noted in #8109 , the cash flow report is mostly incorrect. This is because it is impossible to correctly guess how all accounts in the chart of accounts should affect the cash flow report.

To fix this, the user will need to give hints to ERPNext about where accounts belong in the report.

This PR introduces two new doctypes - Cash Flow Mapper and Cash Flow Mapping, which give users the ability to customise cash flow reports.

#### Three of them by default representing the three sections in cash flow statement
cash-flow-mapper-2

#### Example Mapper for 'Operating Activities'
cash-flow-mapper-4

#### ...For 'Investing Activities'
cash-flow-mapper-5

#### ...And for 'Financing Activities'
cash-flow-mapper-6

`Cash Flow Mapper`s contain links to `Cash Flow Mapping` documents which represent a line in the cash flow report. For example, take a look at the `Cash Flow Mapper` for 'Investing Activities' which has  a `Cash Flow Mapping` called 'Property, plant and equipment'. The Cash Flow Mapping document is shown below:
screenshot-2018-2-16 property plant and equipment

ERPNext will calculate the effect of all the accounts in the above Cash Flow Mapping and display it in the report using the text in the `label` field.

final-cash-flow

If you want the line to appear as 'Fixed assets purchased', simply change the label field to 'Fixed assets purchased', save and refresh your cash flow report.

To use custom cash flow reports, you will need to activate it in Accounts Settings

### CODE REVIEW

File: erpnext/patches/v10_0/add_default_cash_flow_mappers.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
+# Copyright (c) 2017, Frappe and Contributors
+# License: GNU General Public License v3. See license.txt
+
+from __future__ import unicode_literals
+import frappe
+
+
+def execute():
+    mappers = [
```
Comment (nabinhait) [inline erpnext/patches/v10_0/add_default_cash_flow_mappers.py]: These default mappers will also be initiated on fresh install.

**Review by nabinhait — CHANGES_REQUESTED** [review]

File: erpnext/patches/v10_0/add_default_cash_flow_mappers.py (hunk 1 above)
Comment (tundebabzy) [reply erpnext/patches/v10_0/add_default_cash_flow_mappers.py]: I forgot. Thanks

**Comment by tundebabzy** [comment]
done

**Review by nabinhait — APPROVED** [review]

**Comment by mrodriguez614** [comment]
Hi,

i follw the same step. However, the cashflow mapping i created is not reflecting on the cash flow report.

**Comment by hballan** [comment]
Hello,

I also followed the same steps with no success. The cashflow statement is not reflecting any of the custom mappings in the cash flow mapper.

**Pull request merged**